File: src/ios/FreshchatPlugin.m
//
//  FreshchatPlugin.m
//  FreshchatSDK
//
//  Copyright(c) 2016 Freshdesk. All rights reserved.

#import<Cordova/CDV.h>


#import "FreshchatPlugin.h"
#import "FreshchatSDK/FreshchatSDK.h"

#define SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(v)  ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] != NSOrderedAscending)

@implementation FreshchatPlugin:CDVPlugin


-(void) callbackToJavascriptWithResult:(CDVPluginResult*)result ForCommand:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

-(void) callbackToJavascriptWithoutResultForCommand:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* emptyResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self callbackToJavascriptWithResult:emptyResult ForCommand:command];
}

-(void) callbackToJavascriptWithoutResultFailureForCommand:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* emptyResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    [self callbackToJavascriptWithResult:emptyResult ForCommand:command];
}

-(void) callbackToJavascriptWithException:(NSException*)e ForCommand:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[e name]];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

-(void) init:(CDVInvokedUrlCommand*)command {
    NSArray* arguments = [command arguments];
    NSDictionary* initParams;
    if(arguments != nil && arguments.count > 0) {
        initParams = [arguments firstObject];
    } else {
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    NSString* domain = [initParams objectForKey:@"domain"];
    NSString* appId = [initParams objectForKey:@"appId"];
    NSString* appKey = [initParams objectForKey:@"appKey"];
    
    FreshchatConfig *config = [[FreshchatConfig alloc]initWithAppID:appId  andAppKey:appKey];
    NSLog(@"Inside init appId:%@ appKey:%@ domain:%@", appId, appKey, domain);
    
    if(domain) {
        NSLog(@"domain value: %@",domain);
        config.domain = domain;
    }
    
    if(initParams [@"cameraCaptureEnabled"]) {
        config.cameraCaptureEnabled = [[initParams objectForKey:@"cameraCaptureEnabled"] boolValue];
    }
     if(initParams [@"gallerySelectionEnabled"]) {
        config.gallerySelectionEnabled = [[initParams objectForKey:@"gallerySelectionEnabled"] boolValue];
    }
    
    // if(initParams [@"displayFAQsAsGrid"]) {
    //     config.displayFAQsAsGrid = [[initParams objectForKey:@"displayFAQsAsGrid"] boolValue];
    // }
    
    //  if(initParams [@"agentAvatarEnabled"]) {
    //     config.agentAvatarEnabled = [[initParams objectForKey:@"agentAvatarEnabled"] boolValue];
    // }
    
    // if(initParams [@"pictureMessagingEnabled"]) {
    //     config.pictureMessagingEnabled = [[initParams objectForKey:@"pictureMessagingEnabled"] boolValue];
    // }

    if(initParams [@"notificationSoundEnabled"]) {
        config.notificationSoundEnabled = [[initParams objectForKey:@"notificationSoundEnabled"] boolValue];
    }
    
    if(initParams [@"teamMemberInfoVisible"]) {
        config.teamMemberInfoVisible = [[initParams objectForKey:@"teamMemberInfoVisible"] boolValue];
    }

    if(initParams [@"showNotificationBanner"]) {
        config.showNotificationBanner = [[initParams objectForKey:@"showNotificationBanner"] boolValue];
    }
    [[Freshchat sharedInstance] initWithConfig:config];
    [self callbackToJavascriptWithoutResultForCommand:command];
}

- (void) showConversations :(CDVInvokedUrlCommand*)command {
        ConversationOptions *options = [ConversationOptions new];
        NSArray* arguments = [command arguments];
        NSDictionary* conversationParams;
        if(arguments != nil && arguments.count > 0) {
            conversationParams = [arguments firstObject];
            NSMutableArray *tagsList = [NSMutableArray array];
            NSArray* tags = [conversationParams objectForKey:@"tags"];
            if(tags != nil && tags.count > 0) {
                for(int i=0; i<tags.count; i++) {
                    [tagsList addObject:[tags objectAtIndex:i]];
                }
                NSString* title = [conversationParams objectForKey:@"filteredViewTitle"];
                [options filterByTags:tagsList withTitle:title];
            }
            [[Freshchat sharedInstance] showConversations:[self viewController] withOptions: options];
        } else {
            [[Freshchat sharedInstance] showConversations:[self viewController]];
        }
        [self callbackToJavascriptWithoutResultForCommand:command];
    }

- (void) showFAQs :(CDVInvokedUrlCommand*)command {
    NSArray* arguments = [command arguments];
    NSDictionary* faqParams;
    if(arguments != nil && arguments.count > 0) {
        faqParams = [arguments firstObject];
        FAQOptions *options = [FAQOptions new];
        
        if(faqParams [@"showFaqCategoriesAsGrid"]) {
            options.showFaqCategoriesAsGrid = [[faqParams objectForKey:@"showFaqCategoriesAsGrid"] boolValue];
        }
        if(faqParams [@"showContactUsOnFaqScreens"]) {
            options.showContactUsOnFaqScreens = [[faqParams objectForKey:@"showContactUsOnFaqScreens"] boolValue];
        }
        if(faqParams [@"showContactUsOnAppBar"]) {
            options.showContactUsOnAppBar = [[faqParams objectForKey:@"showContactUsOnAppBar"] boolValue];
        }
        NSMutableArray *tagsList = [NSMutableArray array];
        NSArray* tags = [faqParams objectForKey:@"tags"];
        NSString* articleType = [faqParams objectForKey:@"articleType"];
        if(tags != nil && tags.count > 0) {
            for(int i=0; i<tags.count; i++) {
                [tagsList addObject:[tags objectAtIndex:i]];
            }
            NSString* title = [faqParams objectForKey:@"filteredViewTitle"];
            if( [articleType isEqualToString:@"category"]) {
                [options filterByTags:tagsList withTitle:title andType: CATEGORY];
            } else {
                [options filterByTags:tagsList withTitle:title andType: ARTICLE];
            }
        }
        NSMutableArray *contactusTagsList = [NSMutableArray array];
        NSArray* contactusTags = [faqParams objectForKey:@"contactusTags"];
        if(contactusTags != nil && contactusTags.count > 0) {
            for(int i=0; i<contactusTags.count; i++) {
                [contactusTagsList addObject:[contactusTags objectAtIndex:i]];
            }
            NSString* contactusTitle = [faqParams objectForKey:@"contactusFilterTitle"];
            [options filterContactUsByTags:contactusTagsList withTitle:contactusTitle];
        }
        [[Freshchat sharedInstance]showFAQs:[self viewController] withOptions:options];
    } else {
        [[Freshchat sharedInstance]showFAQs:[self viewController]];
    }
    [self callbackToJavascriptWithoutResultForCommand:command];
}

- (void) clearUserData : (CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        [[Freshchat sharedInstance]resetUserWithCompletion:^{
            [self callbackToJavascriptWithoutResultForCommand:command];
        }];
    }];
}

- (void) unreadCount :(CDVInvokedUrlCommand*)command {
    [[Freshchat sharedInstance] unreadCountWithCompletion:^(NSInteger unreadCount) {
        NSLog(@" The unread count value is: %ld", unreadCount);
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)unreadCount];
        [self callbackToJavascriptWithResult:result ForCommand:command];
    }];
}

- (void) updateUser :(CDVInvokedUrlCommand*)command {
    NSArray* arguments = [command arguments];
    NSDictionary* args;
    if(arguments != nil && arguments.count > 0) {
        args = [arguments firstObject];
    } else {
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    FreshchatUser *user = [FreshchatUser sharedInstance];

    if([args objectForKey:@"firstName"] != nil) {
        user.firstName = [args objectForKey:@"firstName"];
    }
    
    if([args objectForKey:@"lastName"] != nil) {
        user.lastName = [args objectForKey:@"lastName"];
    }
    if([args objectForKey:@"email"] != nil) {
        user.email = [args objectForKey:@"email"];
    }
    if([args objectForKey:@"countryCode"] != nil) {
        user.phoneCountryCode = [args objectForKey:@"countryCode"];
    }
    if([args objectForKey:@"phoneNumber"] != nil) {
        user.phoneNumber = [args objectForKey:@"phoneNumber"];
    }
    [self.commandDelegate runInBackground:^{
        [[Freshchat sharedInstance] setUser:user];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }];  
}

- (void) updateUserProperties :(CDVInvokedUrlCommand*)command {

    NSArray* arguments = [command arguments];
    NSDictionary* args;
    if(arguments != nil && arguments.count > 0) {
        args = [arguments firstObject];
    } else {
        [self callbackToJavascriptWithoutResultFailureForCommand:command];
    }

    NSArray *arrayOfKeys = [args allKeys];
    NSArray *arrayOfValues = [args allValues];

    NSString *key;
    NSString *value;

        for(int i=0; i<arrayOfKeys.count; i++) {
            key = [arrayOfKeys objectAtIndex:i];
            value = [arrayOfValues objectAtIndex:i];
            NSLog(@" The userMeta key is: %@ value is: %@", key,value);
            [[Freshchat sharedInstance] setUserPropertyforKey:key withValue:value];
        }
    [self callbackToJavascriptWithoutResultForCommand:command];
}

- (void) getVersionName :(CDVInvokedUrlCommand*)command {
    NSString* versionNumber = [Freshchat SDKVersion];
    NSLog(@"Freshchat version: %@", versionNumber);
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)versionNumber];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

- (void) updatePushNotificationToken :(CDVInvokedUrlCommand*)command {
    NSArray* arguments = [command arguments];
    NSData* devToken;
    if(arguments != nil && arguments.count > 0) {
        devToken = [arguments firstObject];
    } else {
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    NSLog(@"Registration token value: %@", devToken);
    [self.commandDelegate runInBackground:^{
        [[Freshchat sharedInstance] setPushRegistrationToken:devToken];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }];
    NSLog(@"Registration token has been updated");
}

- (void) _isFreshchatPushNotification :(CDVInvokedUrlCommand*)command {
    NSLog(@"checking if freshchat push notification");
    NSArray* arguments = [command arguments];
    NSDictionary* info;
    if(arguments != nil && arguments.count > 0) {
        info = [arguments firstObject];
    } else {
       [self callbackToJavascriptWithoutResultForCommand:command];
    }
    if ([[Freshchat sharedInstance]isFreshchatNotification:info]) {
        NSLog(@"It is a freshchat notification");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)1];
        [self callbackToJavascriptWithResult:result ForCommand:command];
    }
}

- (void) handlePushNotification : (CDVInvokedUrlCommand*)command {
    NSLog(@"Received a freshchat push notification");
    NSArray* arguments = [command arguments];
    NSDictionary* info = [arguments firstObject];
    [self.commandDelegate runInBackground:^{
        [[Freshchat sharedInstance]handleRemoteNotification:info andAppstate:[UIApplication sharedApplication].applicationState];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }];
}
- (void) sendMessage : (CDVInvokedUrlCommand*)command {
    NSLog(@"Send Message called");

    NSArray* arguments = [command arguments];
    NSDictionary* args;
     if(arguments != nil && arguments.count > 0) {
        args = [arguments firstObject];
    } else {
        NSLog(@"Please provide tag and message field in object to send message");
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    if([args objectForKey:@"tag"] != nil && [args objectForKey:@"message"] != nil) {
         NSString* tag = [args objectForKey:@"tag"];
         NSString* message = [args objectForKey:@"message"];
            FreshchatMessage *userMessage = [[FreshchatMessage alloc] initWithMessage:message andTag:tag];
            [self.commandDelegate runInBackground:^{
            [[Freshchat sharedInstance] sendMessage:userMessage];
            [self callbackToJavascriptWithoutResultForCommand:command];
    }];
    }
    else{
        NSLog(@"Please provide tag and message field in object to send message");
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    
 
}

- (void) getRestoreID : (CDVInvokedUrlCommand*)command {
    NSLog(@"getRestoreID called");
    NSString* restoreID = [FreshchatUser sharedInstance].restoreID;
    NSLog(@"Freshchat restoreID: %@", restoreID);
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:restoreID];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

- (void) identifyUser : (CDVInvokedUrlCommand*)command {
    NSLog(@"identifyUser called");
    
    NSArray* arguments = [command arguments];
    NSDictionary* args;
    if(arguments != nil && arguments.count > 0) {
        args = [arguments firstObject];
    } else {
        NSLog(@"Please provide externalId and restoreId in object to identifyUser");
        [self callbackToJavascriptWithoutResultForCommand:command];
    }
    
    NSString* externalId = [args objectForKey:@"externalId"];
    NSString* restoreId = [args objectForKey:@"restoreId"];
    
    if([restoreId isEqual:[NSNull null]]) {
        restoreId = nil;
    }

    [self.commandDelegate runInBackground:^{
        [[Freshchat sharedInstance] identifyUserWithExternalID:externalId restoreID:restoreId];
        [self callbackToJavascriptWithoutResultForCommand:command];
    }];
}
-(void) registerRestoreIdNotification:(CDVInvokedUrlCommand*)command {
   [[NSNotificationCenter defaultCenter] addObserverForName:FRESHCHAT_USER_RESTORE_ID_GENERATED object:nil queue:nil usingBlock:^(NSNotification *note) {
      [self didReceiveNotification:note :command];
    }];
}
-(void) unregisterRestoreIdNotification:(CDVInvokedUrlCommand*)command{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:FRESHCHAT_USER_RESTORE_ID_GENERATED object:nil];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"sucessFully Unsubscrbed unread count listener "];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

-(void) unreadCountlistenerRegister:(CDVInvokedUrlCommand*)command {
    [[NSNotificationCenter defaultCenter]addObserverForName:FRESHCHAT_UNREAD_MESSAGE_COUNT_CHANGED object:nil queue:nil usingBlock:^(NSNotification *note) {
       [[Freshchat sharedInstance]unreadCountWithCompletion:^(NSInteger count) {
            NSLog(@"your unread count : %d", (int)count);
             CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)count];
            [self callbackToJavascriptWithResult:result ForCommand:command];
       }]; 
}];
}
-(void) unreadCountlistenerUnregister:(CDVInvokedUrlCommand*)command{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:FRESHCHAT_UNREAD_MESSAGE_COUNT_CHANGED object:nil];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"sucessFully Unsubscrbed  unread count listener"];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}

- (void) didReceiveNotification :(NSNotification*)notification :(CDVInvokedUrlCommand*)command {
    NSString* restoreID = [FreshchatUser sharedInstance].restoreID;
    NSString* externalID = [FreshchatUser sharedInstance].externalID;
    NSLog(@"Freshchat restoreID: %@", restoreID);
    NSMutableDictionary* param =[[NSMutableDictionary alloc]init];
    [param setValue:restoreID forKey:@"restoreId"];
    [param setValue:externalID forKey:@"externalId"];
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:param];
    [self callbackToJavascriptWithResult:result ForCommand:command];
}



@end

File: src/android/freshchatPlugin.java
//
//  HotlinePlugin.java
//
//  Copyright (c) 2014 Freshdesk. All rights reserved.


package com.freshdesk.freshchat.android;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.ConnectionResult;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;

import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import java.util.Iterator;
import java.util.HashMap;
import java.util.Map;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;
import android.os.Bundle;
import android.widget.Toast;

import com.freshchat.consumer.sdk.*;
import com.google.android.gms.gcm.GoogleCloudMessaging;
import com.google.android.gms.iid.InstanceID;
import java.util.ArrayList;
import java.util.List;





public class freshchatPlugin extends CordovaPlugin {

    private boolean isInitialized = false;
    private FreshchatConfig freshchatConfig;
    private FaqOptions faqOptions;
    private ConversationOptions conversationOptions;
    private Context cordovaContext;
    private static final int PLAY_SERVICES_RESOLUTION_REQUEST = 3458;
    private static final String LOG_TAG = "Freshchat";
    private FreshchatUser freshchatUser;
    private FreshchatMessage freshchatMessage;
    private Map<String, String> userMeta;
    private Bundle bundle;

    private CallbackContext resCallbackContext;
    private CallbackContext unreadResCallbackContext;
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        cordovaContext = cordova.getActivity().getApplicationContext();
    }

    BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
            public void onReceive(Context context, Intent intent) {
            PluginResult result;
            String restoreId = Freshchat.getInstance(cordovaContext).getUser().getRestoreId();
            String externalId = Freshchat.getInstance(cordovaContext).getUser().getExternalId();
            try{
                JSONObject parameter = new JSONObject();
                parameter.put("restoreId", restoreId);
                parameter.put("externalId", externalId);
                result = new PluginResult(PluginResult.Status.OK,parameter);
                result.setKeepCallback(false);
                if (resCallbackContext != null) {
                    resCallbackContext.sendPluginResult(result);
                    resCallbackContext = null;
                }
            }catch(Exception e){
                Log.e(LOG_TAG, e.toString());
            }

        }
    };
    BroadcastReceiver unreadCountChangeReceiver = new BroadcastReceiver() {
    @Override
    public void onReceive(Context context, Intent intent) {
        Freshchat.getInstance(cordovaContext).getUnreadCountAsync(new UnreadCountCallback() {
            @Override
            public void onResult(FreshchatCallbackStatus freshchatCallbackStatus, int unreadCount) {
                PluginResult result = new PluginResult(PluginResult.Status.OK,Integer.toString(unreadCount));
                result.setKeepCallback(false);
                if (unreadResCallbackContext != null) {
                    unreadResCallbackContext.sendPluginResult(result);
                    unreadResCallbackContext = null;
                }
            }
        });
    }
};

    public Bundle jsonToBundle(JSONObject jsonObject) throws JSONException {
        Bundle bundle = new Bundle();
        if(jsonObject == null) {
            return bundle;
        }
        Iterator iterator = jsonObject.keys();
        while(iterator.hasNext()){
            String key = (String)iterator.next();
            String value = jsonObject.getString(key);
            bundle.putString(key,value);
        }
        return bundle;
     }
    
    @Override
    public boolean execute(String action, JSONArray args,final CallbackContext callbackContext) throws JSONException {

        try {
                if(action.equals("init")) {
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide parameters for initializing Freshchat");
                        return false;
                    }
                    JSONObject initArgs = new JSONObject(args.getString(0));
                    Log.d(LOG_TAG,"inside init call");
                    String appId = initArgs.getString("appId");
                    String appKey = initArgs.getString("appKey");

                    freshchatConfig = new FreshchatConfig(appId,appKey);

                    if(initArgs.has("domain")) {
                        freshchatConfig.setDomain(initArgs.getString("domain"));
                    }
                    if(initArgs.has("cameraCaptureEnabled")) {
                        freshchatConfig.setCameraCaptureEnabled(initArgs.getBoolean("cameraCaptureEnabled"));
                    }
                    if(initArgs.has("gallerySelectionEnabled")){

                        freshchatConfig.setGallerySelectionEnabled(initArgs.getBoolean("gallerySelectionEnabled"));
                    }
                     if(initArgs.has("teamMemberInfoVisible")){

                        freshchatConfig.setTeamMemberInfoVisible(initArgs.getBoolean("teamMemberInfoVisible"));
                    }
                    Freshchat.getInstance(cordovaContext).init(freshchatConfig);
                    callbackContext.success();
                    this.isInitialized = true;
                    return true;
                }

                if(action.equals("showFAQs")) {
                    Log.d(LOG_TAG,"Show FAQs has been called");
                    if(args.length() == 0) {
                        Freshchat.showFAQs(cordovaContext);
                        return true;
                    }
                    JSONObject faqArgs = new JSONObject(args.getString(0));
                    faqOptions = new FaqOptions();
                    if(faqArgs.has("showFaqCategoriesAsGrid")) {
                        faqOptions.showFaqCategoriesAsGrid(faqArgs.getBoolean("showFaqCategoriesAsGrid"));
                    }
                    if(faqArgs.has("showContactUsOnAppBar")) {
                        faqOptions.showContactUsOnAppBar(faqArgs.getBoolean("showContactUsOnAppBar"));
                    }
                    if(faqArgs.has("showContactUsOnFaqScreens")) {
                        faqOptions.showContactUsOnFaqScreens(faqArgs.getBoolean("showContactUsOnFaqScreens"));
                    }
                    if(faqArgs.has("showContactUsOnFaqNotHelpful")) {
                        faqOptions.showContactUsOnFaqNotHelpful(faqArgs.getBoolean("showContactUsOnFaqNotHelpful"));
                    }

                    List<String> tagsList = new ArrayList<String>();
                    if(faqArgs.optJSONArray("tags") != null) {
                        JSONArray tags = faqArgs.getJSONArray("tags");
                        for (int i = 0; i < tags.length(); i++) {
                            tagsList.add(tags.getString(i));
                        }
                        String title = faqArgs.getString("filteredViewTitle");
                        if(faqArgs.getString("articleType").equals("category")){
                            faqOptions.filterByTags(tagsList, title, FaqOptions.FilterType.CATEGORY);
                        } else {
                                faqOptions.filterByTags(tagsList, title, FaqOptions.FilterType.ARTICLE);
                        }
                        List<String> contactusTagsList = new ArrayList<String>();
                        if(faqArgs.optJSONArray("contactusTags") != null) {
                            JSONArray contactusTags = faqArgs.getJSONArray("contactusTags");
                            for (int i = 0; i < contactusTags.length(); i++) {
                                contactusTagsList.add(contactusTags.getString(i));
                            }
                            title = faqArgs.getString("contactusFilterTitle");
                            faqOptions.filterContactUsByTags(contactusTagsList, title);
                        }
                        Freshchat.showFAQs(cordovaContext, faqOptions);
                    } else {
                        Freshchat.showFAQs(cordovaContext);
                    }
                    callbackContext.success();
                    return true;
                }

                if(action.equals("showConversations")) {
                    Log.d(LOG_TAG,"show Conversations has been called");
                    if(args.length() == 0) {
                        Freshchat.showConversations(cordovaContext);
                        return true;
                    }
                    JSONObject conversationArgs = new JSONObject(args.getString(0));
                    conversationOptions = new ConversationOptions();
                    List<String> tagsList = new ArrayList<String>();
                    if(conversationArgs.optJSONArray("tags") != null) {
                        JSONArray tags = conversationArgs.getJSONArray("tags");
                        for (int i = 0; i < tags.length(); i++) {
                            tagsList.add(tags.getString(i));
                        }
                        String title = conversationArgs.getString("filteredViewTitle");
                        conversationOptions.filterByTags(tagsList, title);
                    }
                    Freshchat.showConversations(cordovaContext, conversationOptions);
                    callbackContext.success();
                    return true;
                }

                if(action.equals("clearUserData")) {
                    Log.d(LOG_TAG,"inside clearUserData");
                     Freshchat.resetUser(cordovaContext);
                     callbackContext.success();
                    return true;
                }

                if(action.equals("updateUser")) {
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide parameters to update a user");
                        return false;
                    }
                    JSONObject jsonArgs = new JSONObject(args.getString(0));

                    freshchatUser=Freshchat.getInstance(cordovaContext).getUser();

                    if(jsonArgs.has("firstName")) {
                        freshchatUser.setFirstName(jsonArgs.getString("firstName"));
                    }
                    if(jsonArgs.has("lastName")) {
                        freshchatUser.setLastName(jsonArgs.getString("lastName"));
                    }
                    if(jsonArgs.has("email")) {
                        freshchatUser.setEmail(jsonArgs.getString("email"));
                    }
                    if(jsonArgs.has("countryCode") && jsonArgs.has("phoneNumber")) {
                        freshchatUser.setPhone(jsonArgs.getString("countryCode"),jsonArgs.getString("phoneNumber"));
                    }
                    Freshchat.getInstance(cordovaContext).setUser(freshchatUser);
                    callbackContext.success();
                    return true;
                }

                if(action.equals("updateUserProperties")) {

                        if(args.length() == 0) {
                            Log.e(LOG_TAG,"Please provide user properties to update the user");
                            return false;
                        }

                        JSONObject metadata = new JSONObject(args.getString(0));
                        userMeta = new HashMap<String, String>();
                        Iterator<String> keys  = metadata.keys();

                        while(keys.hasNext()) {
                                String key = keys.next();
                                userMeta.put(key, metadata.getString(key));
                        }
                        Freshchat.getInstance(cordovaContext).setUserProperties(userMeta);
                        callbackContext.success();
                        return true;
                }

                if(action.equals("unreadCount")) {
                    Log.d(LOG_TAG," unreadCount has been called");
                    if(args.length() == 0) {
                    Freshchat.getInstance(cordovaContext).getUnreadCountAsync(new UnreadCountCallback() {
                        @Override
                        public void onResult(FreshchatCallbackStatus freshchatCallbackStatus, int count)
                        {
                            if (freshchatCallbackStatus == FreshchatCallbackStatus.STATUS_SUCCESS) {
                                Log.i(LOG_TAG,"unreadcount is :"+count);
                                callbackContext.success(count);
                            }
                            else{
                                callbackContext.error(freshchatCallbackStatus.toString());
                            }
                                
                        }
                    });
                    return true;
                    }
                    JSONObject unreadCountArgs = new JSONObject(args.getString(0));
                    List<String> tagsList = new ArrayList<String>();
                    if(unreadCountArgs.has("tag")) {
                        tagsList.add(unreadCountArgs.getString("tag"));
                       Freshchat.getInstance(cordovaContext).getUnreadCountAsync(new UnreadCountCallback() {
                        @Override
                        public void onResult(FreshchatCallbackStatus freshchatCallbackStatus, int count)
                        {
                            if (freshchatCallbackStatus == FreshchatCallbackStatus.STATUS_SUCCESS) {
                                Log.i(LOG_TAG,"unreadcount is :"+count);
                                callbackContext.success(count);
                            }
                            else{
                                callbackContext.error(freshchatCallbackStatus.toString());
                            }
                                
                            }
                    },tagsList);
                    }
                    return true;
                  
                }
                if(action.equals("unreadCountlistenerRegister")){
                    IntentFilter intentFilter = new IntentFilter(Freshchat.FRESHCHAT_UNREAD_MESSAGE_COUNT_CHANGED);
                    LocalBroadcastManager.getInstance(cordovaContext).registerReceiver(unreadCountChangeReceiver, intentFilter);
                    PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
                    result.setKeepCallback(true);
                    unreadResCallbackContext = callbackContext;
                    callbackContext.sendPluginResult(result);
                    return true;
                }
                 if(action.equals("unreadCountlistenerUnregister")){
                    if(unreadResCallbackContext != null){
                        LocalBroadcastManager.getInstance(cordovaContext).unregisterReceiver(unreadCountChangeReceiver);
                        callbackContext.success("success");
                    }
                    
                    return true;
                }

                if(action.equals("getVersionName")) {
                    Log.d(LOG_TAG,"version number called");
                    int versionNumber = Freshchat.getInstance(cordovaContext).getSDKVersionCode();
                    callbackContext.success(versionNumber);
                    return true;
                }

                if(action.equals("_isFreshchatPushNotification")) {
                    Log.d(LOG_TAG,"check if a particular push notificaiton is a freshchat push notification or not");
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide the notification payload to be verified ");
                        return false;
                    }
                    JSONObject jsonArgs = new JSONObject(args.getString(0));
                    Bundle bundle = jsonToBundle(jsonArgs);
                    if(Freshchat.getInstance(cordovaContext).isFreshchatNotification(bundle)) {
                        callbackContext.success(1);
                    }
                    else {
                        callbackContext.success(0);
                    }
                    return true;
                }

                if(action.equals("handlePushNotification")) {
                    Log.d(LOG_TAG,"Handling Push Notification!");
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide parameters for initializing FreshChat");
                        return false;
                    }
                    JSONObject jsonArgs = new JSONObject(args.getString(0));
                    bundle = jsonToBundle(jsonArgs);
                    cordova.getThreadPool().execute( new Runnable() {
                       public void run() {
                            Freshchat.getInstance(cordovaContext).handleFcmMessage(cordovaContext,bundle);
                            callbackContext.success();
                       }
                    });

                    return true;
                }
                if(action.equals("getRestoreID")) {
                    Log.d(LOG_TAG,"getRestoreID called");
                    freshchatUser= Freshchat.getInstance(cordovaContext).getUser();
                    String restoreID = freshchatUser.getRestoreId();
                    Log.d(LOG_TAG,restoreID);
                    callbackContext.success(restoreID);
                    return true;
                }
                if(action.equals("registerRestoreIdNotification")){
                     Log.d(LOG_TAG,"listenRestoreId called");
                    IntentFilter intentFilter = new IntentFilter(Freshchat.FRESHCHAT_USER_RESTORE_ID_GENERATED);
                    LocalBroadcastManager.getInstance(cordovaContext).registerReceiver(broadcastReceiver, intentFilter);
                    PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
                    result.setKeepCallback(true);
                    resCallbackContext = callbackContext;
                    callbackContext.sendPluginResult(result);
                    return true;
                }
                if(action.equals("unregisterRestoreIdNotification")){
                    if(resCallbackContext != null){
                        LocalBroadcastManager.getInstance(cordovaContext).unregisterReceiver(broadcastReceiver);
                    }
                    
                    return true;
                }
                if (action.equals("identifyUser")) {
                    if (args.length() == 0) {
                        String errorMsg = "Please provide parameters to identifyUser";
                        Log.e(LOG_TAG, errorMsg);
                        callbackContext.error("identifyUser failed - " + errorMsg);
                        return false;
                    }
                    try {
                        JSONObject jsonArgs = new JSONObject(args.getString(0));
                        String externalId = null;
                        String restoreId = null;

                        if (jsonArgs.has("externalId")) {
                            externalId = jsonArgs.getString("externalId");
                        }

                        if (jsonArgs.has("restoreId")) {
                            restoreId = jsonArgs.getString("restoreId");
                        }

                        Freshchat.getInstance(cordovaContext).identifyUser(externalId, restoreId);
                        callbackContext.success("identifyUser successfully");
                        return true;
                    } catch (Exception e) {
                        callbackContext.error("identifyUser failed - " + e.toString());
                        return false;
                    }
                }

                if(action.equals("sendMessage")) {
                        if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide parameters to send Message");
                        return false;
                    }
                    JSONObject jsonArgs = new JSONObject(args.getString(0));
                    freshchatMessage = new FreshchatMessage();
                    if(jsonArgs.has("tag")&&  jsonArgs.has("message"))  {
                        freshchatMessage.setTag(jsonArgs.getString("tag")).setMessage(jsonArgs.getString("message"));
                        Freshchat.getInstance(cordovaContext).sendMessage(cordovaContext, freshchatMessage);
                        callbackContext.success("message sent.");
                    }else{
                        callbackContext.error("Failed to send");
                    }

                    return true;
                }

                if(action.equals("updatePushNotificationToken")) {
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide the token to register");
                        return false;
                    }
                    String token = args.getString(0);
                    Log.i(LOG_TAG,"update GCM registration has been called");
                    Freshchat.getInstance(cordovaContext).setPushRegistrationToken(token);
                    return true;
                }

                if(action.equals("updateAndroidNotificationProperties")) {
                    Log.d(LOG_TAG,"setting notification properties!");
                    if(args.length() == 0) {
                        Log.e(LOG_TAG,"Please provide parameters to update a user");
                        return false;
                    }
                    JSONObject jsonArgs = new JSONObject(args.getString(0));
                    FreshchatNotificationConfig notificationConfig = new FreshchatNotificationConfig();
                    
                    if(jsonArgs.has("notificationSoundEnabled")) {
                        notificationConfig.setNotificationSoundEnabled(jsonArgs.getBoolean("notificationSoundEnabled"));
                    }
                    if(jsonArgs.has("smallIcon")) {
                        String iconName = jsonArgs.getString("smallIcon");
                        int iconId = cordovaContext.getResources().getIdentifier(iconName, "drawable", cordovaContext.getPackageName());
                        notificationConfig.setSmallIcon(iconId);
                    }
                    if(jsonArgs.has("largeIcon")) {
                        String iconName = jsonArgs.getString("largeIcon");
                        int iconId = cordovaContext.getResources().getIdentifier(iconName, "drawable", cordovaContext.getPackageName());
                        notificationConfig.setLargeIcon(iconId);
                    }
                    if(jsonArgs.has("launchActivityOnFinish")) {
                        notificationConfig.launchActivityOnFinish(jsonArgs.getString("launchActivityOnFinish"));
                    }
                    if(jsonArgs.has("notificationPriority")) {
                        notificationConfig.setPriority(jsonArgs.getInt("notificationPriority"));
                    }
                    Freshchat.getInstance(cordovaContext).setNotificationConfig(notificationConfig);
                    return true;
                }
                Log.d(LOG_TAG,"action does not have a function to match it:"+action);

            } catch (Exception e) {
                Log.e(LOG_TAG,"exception while perfroming action:"+action,e);
                callbackContext.error("exception while performing action"+action);
            }
        return true;
    }



}
